### PR TITLE
feat: Add rank transformer

### PR DESCRIPTION
**Pull request by jamesdshinner**
### Description
Adds a transformer to go from a score to a rank, e.g. for score [1.0, 1.5, 2.0] the rank would be [3, 2, 1].

In cases of ties in score, the original order is used to decide the rank for consistency and reproducibility.

### Keras Layer Checklist
Verify that:
- [x] The new Keras layer extends [BaseLayer](../blob/main/src/kamae/tensorflow/layers/base.py)
- [x] The `_call` method has been implemented in the new layer.
- [x] The `compatible_dtypes` property is defined in the new layer.
- [x] The new layer is decorated with `@tf.keras.utils.register_keras_serializable(package=kamae.__name__)`.
- [x] The new layer takes a `name`, `input_dtype`, and `output_dtype` as arguments to the constructor and that this is passed to the super constructor.
- [x] The Keras layer is serializable. I have implemented the `get_config` method.
- [x] There are unit tests of the new layer. 
- [x] There is a specific test of layer serialisation added [here](../blob/main/tests/kamae/tensorflow/test_layer_serialisation.py).
- [x] The new layer is imported in the [__init__.py](../blob/main/src/kamae/tensorflow/layers/__init__.py) file in the `layers` directory.

### Spark Transformer/Estimator Checklist
Verify that:
- [x] The new Spark Transformer extends [BaseTransformer](../blob/main/src/kamae/spark/transformers/base.py).
- [x] If the new transform needs a fit method, a Spark Estimator has been implemented that extends [BaseEstimator](../blob/main/src/kamae/spark/estimators/base.py).
- [x] The instructions in the above docs page have been followed for the `__init__` and `setParams` methods.
- [x] The transformer uses one of the input/output mixin classes from [base.py](../blob/main/src/kamae/spark/params/base.py).
- [x] If the new transformer requires more parameters that would need to be serialised to the Spark ML pipeline, there is a implemented parameter class by extending the `Params` class [here](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.ml.param.Params.html).
- [x] The `compatible_dtypes` property has been implemented to specify the input/output data types that my transformer/estimator supports.
- [x] A Keras subclassed layer is returned in the transformer's `get_tf_layer` method.
- [x] There are unit tests of the new transform. In particular, there are parity tests between the Spark and Keras implementations.
- [x] The new transformer/estimator is imported in the [__init__.py](../blob/main/src/kamae/spark/transformers/__init__.py) file in the `transformers`/`estimators` directory.

Finally, please verify that:
- [x] There is a new entry (alphabetical order) in the README table describing the new layer/transformer

### CODE REVIEW

File: src/kamae/spark/transformers/list_rank.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+        :returns: Tensorflow keras layer with name equal to the layerName parameter that
+         performs a ranking operation.
+        """
+        return ListRankLayer(
+            name=self.getLayerName(),
+            input_dtype=self.getInputTFDtype(),
+            output_dtype=self.getOutputTFDtype(),
+            sort_order=self.getSortOrder(),
+            axis=1,
```
Comment (georyetti) [inline src/kamae/spark/transformers/list_rank.py:142]: The default is `1` so I think we can remove this

File: src/kamae/tensorflow/layers/list_rank.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+    Example: calculate the rank of items within a query, given the score.
+    """
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        input_dtype: Optional[str] = None,
+        output_dtype: Optional[str] = None,
+        sort_order: str = "asc",
```
Comment (georyetti) [inline src/kamae/tensorflow/layers/list_rank.py:39]: Same on default

File: tests/kamae/tensorflow/layers/test_list_rank.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        input_dtype,
+        output_dtype,
+    ):
+        # when
+        name = "listwise_rank_test"
+        layer = ListRankLayer(
+            name=name,
+            input_dtype=input_dtype,
+            output_dtype=output_dtype,
+        )
```
Comment (georyetti) [inline tests/kamae/tensorflow/layers/test_list_rank.py:76]: Can we test for diff axis in the tf side?

File: src/kamae/tensorflow/layers/list_rank.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    def _call(self, inputs: Iterable[Tensor], **kwargs) -> Tensor:
+        """
+        Calculate the rank.
+
+        :param inputs: The iterable tensor for the feature.
+        :returns: The new tensor result column.
+        """
+        # If desc order, multiply tensor by -1
+        if self.sort_order == "desc":
+            inputs = tf.math.multiply(inputs, tf.constant(-1, dtype=inputs.dtype))
```
Comment (georyetti) [inline src/kamae/tensorflow/layers/list_rank.py:90]: argsort has a `direction` parameter, can we not use it?

File: src/kamae/spark/params/shared.py [hunk 5]
```diff
@@ -458,6 +458,83 @@ class ListwiseStatisticsParams(Params):
...
+        return self._set(defaultValue=value)
+
+    def getDefaultValue(self) -> int:
+        """
+        Gets the default value parameter.
+
+        :returns: Int defaults to 0, should a method only have
+        padded data this value will be returned.
+        """
+        return self.getOrDefault(self.defaultValue)
```
Comment (georyetti) [inline src/kamae/spark/params/shared.py:530]: We don't have a default value in ListRank, so I think this needs to go in ListwiseStatistics right?

File: src/kamae/tensorflow/layers/list_rank.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        """
+        Gets the configuration of the layer.
+        Used for saving and loading from a model.
+
+        :returns: Dictionary of the configuration of the layer.
+        """
+        config = super().get_config()
+        config.update(
+            {
+                "axis": self.axis,
```
Comment (georyetti) [inline src/kamae/tensorflow/layers/list_rank.py:110]: We need to add sort order into the config here otherwise it won't serialise it

File: tests/kamae/tensorflow/test_layer_serialisation.py [hunk 7]
```diff
@@ -215,6 +216,7 @@
             },
             False,
         ),
+        (ListRankLayer, [tf.random.normal((1, 2, 3))], None, False),
```
Comment (georyetti) [inline tests/kamae/tensorflow/test_layer_serialisation.py:219]: Can we add some params here to test serialisation? E.g. add axis and sort order instead of None

File: src/kamae/spark/params/shared.py (hunk 5 above)
Comment (jamesdshinner) [reply src/kamae/spark/params/shared.py:530]: We have a default for sort order, so I assumed we needed it for that
